modules/core/src/main/scala/org/tesselation/infrastructure/gossip/GossipDaemon.scala

```diff
@@ -49,44 +51,57 @@ object GossipDaemon {
     private def consumeRumors: F[Unit] =
       Stream
         .fromQueueUnterminated(rumorQueue)
-        .evalMap { batch =>
-          batch.filterA {
-            case (hash, signedRumor) =>
-              for {
-                computedHash <- signedRumor.value.hashF
-                valid = computedHash == hash
-                _ <- if (valid) Applicative[F].unit
-                else
-                  logger.warn(
-                    s"Discarding rumor of type ${signedRumor.value.tpe} with invalid hash. " +
-                      s"Received hash ${hash.show}, actual hash ${computedHash.show}."
-                  )
-              } yield valid
-
-          }
-        }
-        .evalMap { batch =>
-          // TODO: filter out invalid signatures
-          Applicative[F].pure(batch)
-        }
+        .evalMap(validateHash)
+        .evalMap(validateSignature)
         .evalMap(rumorStorage.addRumors)
         .flatMap(batch => Stream.iterable(batch))
-        .parEvalMapUnordered(cfg.maxConcurrentHandlers) {
-          case (hash, signedRumor) =>
-            rumorHandler
-              .run(signedRumor)
-              .getOrElseF {
-                logger.warn(s"Unhandled rumor of type ${signedRumor.value.tpe} with hash ${hash.show}.")
-              }
-              .handleErrorWith { err =>
-                logger.error(err)(
-                  s"Error handling rumor of type ${signedRumor.value.tpe} with hash ${hash.show}."
-                )
-              }
-        }
+        .filter { case (_, signedRumor) => signedRumor.value.origin != nodeId }
+        .parEvalMapUnordered(cfg.maxConcurrentHandlers)(handleRumor)
         .compile
         .drain
 
+    private def validateHash(batch: RumorBatch): F[RumorBatch] =
+      batch.filterA {
+        case (hash, signedRumor) =>
+          for {
+            computedHash <- signedRumor.value.hashF
+            valid = computedHash == hash
+            _ <- if (valid) Applicative[F].unit
+            else
+              logger.warn(
+                s"Discarding rumor of type ${signedRumor.value.tpe} with hash ${hash.show} due to invalid hash. " +
+                  s"Actual hash was ${computedHash.show}."
+              )
+          } yield valid
```

Comment:
it can be simplified to this:

```scala
signedRumor.value.hashF
  .map(_ == hash)
  .ifM(
    Applicative[F].unit,
    logger.warn(...)
  )
```

Reply:
We need to return an `F[Boolean]` from this closure. We could do it like this
```
signedRumor.value.hashF
  .map(_ == hash)
  .ifM(
    Applicative[F].pure(true),
    logger.warn(...) >> Applicative[F].pure(false)
  )
```
but I'm not sure if this is actually better. It's easy to make a mistake this way, because we're not returning the original result of comparison, but arbitrary `true/false` values. What do you think?

Reply:
ah, okay. didn't see that. if so, this would work:
```scala
signedRumor.value.hashF
  .map(_ == hash)
  .flatTap { valid =>
    if (valid) Applicative[F].unit else logger.warn(...)
  }
```
It's actually the same as the for comprehension, just without intermediate names. so up to you :)